### PR TITLE
CMSIS_NN update: 

### PR DESCRIPTION
**Pull request by liangzhen-lai**
adding doxygen documentation,
adding non-square kernels,
adding rouding config options.

### CODE REVIEW

File: CMSIS/DSP/Examples/ARM/arm_nn_examples/gru/arm_nnexamples_gru.cpp [hunk 1]
```diff
@@ -1,29 +1,40 @@
...
-
 /* ----------------------------------------------------------------------
- * Project:      CMSIS-NN
- * Title:        arm_nnexamples_gru.cpp
- * Description:	 Gated Recurrent Unit Example
- *
- * Target Processor: Cortex-M4 and Cortex-M7 cores
- *
- * -------------------------------------------------------------------- */
+* Copyright (C) 2010-2017 ARM Limited. All rights reserved.
```
Comment (JonatanAntoni) [inline CMSIS/DSP/Examples/ARM/arm_nn_examples/gru/arm_nnexamples_gru.cpp]: Please assure all copyright statements are up to date, i.e. this one should become 
> Copyright (C) 2010-2018 Arm Limited. All rights reserved.

File: CMSIS/DSP/Source/NNFunctions/arm_nn_activations.c [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+   * @details
+   * 
+   * This is the direct table look-up approach.
+   *
+   * Assume here the integer part of the fixed-point is <= 3.
+   * More than 3 just not making much sense, makes no difference with
+   * saturation followed by any of these activation functions. 
+   */
+
+void arm_nn_activations_direct_q7(q7_t * data, uint16_t size, uint16_t int_width, arm_nn_activation_type type)
```
Comment (JonatanAntoni) [inline CMSIS/DSP/Source/NNFunctions/arm_nn_activations.c]: Shouldn't this one go into a separate source file named  `arm_nn_activations_direct_q7.c`?

File: CMSIS/DSP/Examples/ARM/arm_nn_examples/nn_test/Ref_Implementations/ref_functions.h [hunk 3]
```diff
@@ -49,6 +50,28 @@ extern    "C"
...
+                                           const q7_t * bias, // bias
+                                           const uint16_t bias_shift, 
+	    								   const uint16_t out_shift, 
+		    							   q7_t * Im_out,    // output image
+                                           const uint16_t dim_im_out_x, // output image dimension x
+				    					   const uint16_t dim_im_out_y, // output image dimension y
+                                           q15_t * bufferA,   //buffer space for input
+                                           q7_t * bufferB //buffer space for output
+        );
+
```
Comment (FabKlein) [inline CMSIS/DSP/Examples/ARM/arm_nn_examples/nn_test/Ref_Implementations/ref_functions.h]: indentation to be fixed

File: CMSIS/DSP/Examples/ARM/arm_nn_examples/nn_test/arm_nnexamples_nn_test.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+#include "arm_math.h"
+    
+#include "arm_nnfunctions.h"
+#include "ref_functions.h"
+
+void initialize_results_q7(q7_t * ref, q7_t * opt, int length)
+{
+  arm_fill_q7(0, ref, length);
+	arm_fill_q7(37, opt, length);
+}
```
Comment (FabKlein) [inline CMSIS/DSP/Examples/ARM/arm_nn_examples/nn_test/arm_nnexamples_nn_test.h]: indentation to be fixed

**Comment by JonatanAntoni** [comment]
I rebased your pull-request and published the result as a new branch features/cmsis_nn_2.
May I ask you to verify the new branch reflects you desired state, please?

**Pull request closed without merging**